ambry-commons/src/test/java/com.github.ambry.commons/BlobIdTest.java

```diff
@@ -13,119 +13,538 @@
-    System.out.println("Blob Id toString: " + blobId);
-    System.out.println("Blob id sizeInBytes: " + blobId.toString().length());
+  /**
+   * Tests deserialization from a composed blobId string (not from deserialization of a blobId in both version
+   * 1 and 2.
+   * @throws Exception Any unexpected exception.
+   */
+  @Test
+  public void testDeserialization() throws Exception {
+    for (Short version : versions) {
+      composedBlobIdAndDeserialize(version);
+    }
   }
 
   /**
-   * Test the parsing of a valid blob ID from a string and stream.
-   * @throws Exception
+   * Tests first serializing a blobId into string, and then deserializing into a blobId object from the string.
+   * The tests are performed on both blob v1 and v2.
+   * @throws Exception Any unexpected exception.
    */
   @Test
-  public void goodIdTest() throws Exception {
-    MockClusterMap clusterMap = new MockClusterMap();
-    short goodVersion = 1;
-    PartitionId goodPartitionId = clusterMap.getWritablePartitionIds().get(0);
-    String goodUUID = UUID.randomUUID().toString();
-    String goodId = buildBlobIdLike(goodVersion, goodPartitionId, goodUUID.length(), goodUUID);
-    List<BlobId> blobIds = new ArrayList<>();
-    blobIds.add(new BlobId(goodId, clusterMap));
-    blobIds.add(new BlobId(getStreamFromBase64(goodId), clusterMap));
-    blobIds.add(new BlobId(getStreamFromBase64(goodId + "EXTRA"), clusterMap));
-    for (BlobId blobId : blobIds) {
-      assertEquals("Wrong partition ID in blob ID: " + blobId, goodPartitionId, blobId.getPartition());
-      assertEquals("Wrong base-64 ID in blob ID: " + blobId, goodId, blobId.getID());
+  public void testSerDes() throws Exception {
+    for (Short version : versions) {
+      serDesBlobId(version);
     }
   }
 
   /**
-   * Test various invalid blob IDs
-   * @throws Exception
+   * Test various invalid blobIds
+   * @throws Exception Any unexpected exception.
    */
   @Test
   public void badIdTest() throws Exception {
-    MockClusterMap clusterMap = new MockClusterMap();
-    List<String> blobIdLikes = new ArrayList<>();
+    for (Short version : versions) {
+      generateAndAssertBadBlobId(version);
+    }
+  }
+
+  /**
+   * Tests blobId long form in both v1 and v2.
+   */
+  @Test
+  public void testBlobIdLongForm() {
+    assertBlobIdLongForm(BLOB_ID_V1);
+    assertBlobIdLongForm(BLOB_ID_V2);
+  }
+
+  /**
+   * Makes a blobId-like string in the same way as used in blobId serialization, and then deserialize from the
+   * string to construct blobId object.
+   * @param version The version of BlobId.
+   * @throws Exception Any unexpected exception.
+   */
+  private void composedBlobIdAndDeserialize(short version) throws Exception {
+    String srcUUID = UUID.randomUUID().toString();
+    String srcBlobIdStr;
+    switch (version) {
+      case BLOB_ID_V1:
+        srcBlobIdStr =
+            buildBlobIdLike(version, null, null, null, null, referencePartitionId, srcUUID.length(), srcUUID);
```

Comment:
since we use instance variables for most part to construct BlobId, may be you can introduce a privateMethod which takes in non-instance variables and version to construct BlobId is the specified version